src/pkg/caendr/caendr/models/datastore/report_entity.py

```diff
@@ -0,0 +1,138 @@
+# Built-ins
+from abc import abstractmethod
+from enum import Enum
+from typing import Union
+
+# Parent class
+from caendr.models.report    import GCPReport
+from caendr.models.datastore import JobEntity, UserOwnedEntity
+
+# Models
+from caendr.models.status    import JobStatus
+from caendr.models.datastore import Container, User
+
+# Services
+from caendr.services.cloud.storage import check_blob_exists
+
+
+
+class ReportEntity(JobEntity, UserOwnedEntity, GCPReport):
+  '''
+    Stores metadata in a Datastore Entity.
+    Manages an entity (metadata) and paths for input/output data
+
+    An Entity subclass that fulfills the Report template class, making it directly usable as a Report in the Job Pipeline.
+
+    Note that the abstract base class GCPReport comes last in the parent class list.
+    This appears to be necessary for the method definitions in the Entity classes
+    to override the abstract methods from GCPReport correctly.
+  '''
+
+  #
+  # Kind
+  #
+
+  @classmethod
+  def get_kind(cls):
+    return cls.kind
+
+
+  #
+  # Instantiation
+  # Route the GCPReport's create & lookup methods to the appropriate Entity methods
+  #
+
+  @classmethod
+  def create(cls, **kwargs):
+    return cls(**kwargs)
+
+  @classmethod
+  def lookup(cls, report_id: str):
+    return cls.get_ds(report_id, silent = False)
+
+
+  #
+  # Data ID
+  # Subclass can designate a property field to use as the data ID
+  #
+
+  # The report field to use as the data ID
+  # Subclass should overwrite this abstract property,
+  # then the chosen prop value will be accessible through the getter below
+  @property
+  @abstractmethod
+  def _data_id_field(self) -> str:
```

Comment:
The data ID is the field I use to group Cloud Run jobs -- different submissions with the same data ID become different executions of the same job.

For the three tools, this is the data hash.  For database operations, this is the operation type (`DROP_AND_POPULATE_STRAINS`, etc.)

See `Report` class for more.